_drafts/building-a-cli-for-firmware-projects.md

```diff
@@ -0,0 +1,514 @@
+  command..."**
+- Scripts have **undocumented** arguments and environment variable overrides
+- Two or more teammates have **written similar scripts** for themselves but a
+  common shared version doesn't exist, or is broken.
+- You check out a 3 month-old revision of your project and there is no longer
+  documentation about how to run (factory builds anyone?).
+
+I could go on and on, but I hope the idea is clear. Building a CLI using Invoke
+helps with all of the above issues and more.
+
+## Getting started with Invoke
+
+Invoke is a "task execution tool & library" that works with Python 2.7 and 3.4+,
+and be installed using `pip`. We'll be using it today to build our project's CLI
+tool.
+
+We'll be using a virtual environment, and I highly recommend them.
+[This guide](https://docs.python-guide.org/dev/virtualenvs/#lower-level-virtualenv)
+is a good starting point. I will be using Python 3.6, which has the `virtualenv`
+command available already.
+
+```
+$ virtualenv venv
+$ source venv/bin/activate
+(venv)
+$ pip install invoke
+```
+
+We can quickly test if it was properly installed by running `invoke` in our
+shell.
+
+```
+(venv)
+$ invoke --list
+Can't find any collection named 'tasks'!
+```
+
+We've confirmed it's installed. To start adding tasks to Invoke, one needs to
+add a `tasks.py` file to the root of a project. No matter where the user is
+within the project, Invoke will search upwards until it finds a `tasks.py` file.
+
+## Using Invoke with the nRF5 SDK
+
+> For purposes of this example, we are going to use the Nordic nRF5 SDK version
+> 15.2.0, which can be downloaded
+> [here](https://developer.nordicsemi.com/nRF5_SDK/nRF5_SDK_v15.x.x/), and
+> within that, the Blinky app located at
+> `nrf5_sdk/examples/peripheral/blinky/pca10056/blank/armgcc/`.
+
+The following is the most basic, but still reasonable, `tasks.py` to allow us to
+easily build our Blinky project:
+
+```python
+# /nrf5_sdk/tasks.py
+
+import os
+
+from invoke import Collection, task
+
+ROOT_DIR = os.path.dirname(__file__)
+BLINKY_DIR = os.path.join("nrf5_sdk", "examples", "peripheral",
+                          "blinky", "pca10056", "blank", "armgcc")
+
+@task
+def build(ctx):
+    """Build the project"""
+    with ctx.cd(PCA10056_BLINKY_ARMGCC_DIR):
+        ctx.run("make build")
+```
+
+This sets up the proper paths, `cd`'s into the directory, and runs `make`.
+
+```
+$ inv --echo build
+cd /Users/tyler/dev/memfault/interrupt/example/invoke-basic/nrf5_sdk/examples/peripheral/blinky/pca10056/blank/armgcc && make -j4
+mkdir _build
+cd _build && mkdir nrf52840_xxaa
+Assembling file: gcc_startup_nrf52840.S
+Compiling file: main.c
+Compiling file: nrf_log_frontend.c
+Compiling file: nrf_log_str_formatter.c
+```
+
+Some of you might immediately think this wrapper or abstraction is unnecessary,
+and maybe it is in the simplest case, but let's dive a bit into the future of
+this build command.
+
+### Adding Parallel Builds
```

Comment:
Some structure suggestions. I'd split the "evolving our workflow" from the "setting the stage portion.
```
## Why Invoke?
### Why you need a CLI
### Why make is not enough
### Why python
## Setting the stage
### Installing invoke
### A basic invoke script
## Iterating on our process
### Adding parallel
### ccache
### etc..
## A more complex example
## Closing
```